File: analysis/scripts/R3_script_for_coverage_plot_MinION_run_B2.R
############### R script for coverage plot

### Set wd
path_setwd <- getwd()
setwd(path_setwd)


# Retrieve command-line arguments
args <- commandArgs(trailingOnly = TRUE)

# Define argument names and default values
arg_names <- c("-m", "-g", "-d", "-a")
arg_values <- setNames(rep(NA, length(arg_names)), arg_names)

# Ensure sufficient arguments are provided
if (length(args) < length(arg_names) * 2) {
  stop("\n❌ Missing required arguments.\n\nUsage:\n  Rscript script.R -m MinION_run_name -g genotyping_method -d analysis_date -a analysis_dr\n\nArguments:\n  -m  MinION run name          (e.g. GAMP_22906B2)\n  -g  Genotyping method        (e.g. clair3_diploid)\n  -d  Analysis date            (format: YYYYMMDD, e.g. 20250511)\n  -a  Path to output directory (e.g. /mnt/c/.../output/)\n")
}

# Parse arguments using vectorized lookup
for (i in seq(1, length(args), by = 2)) {
  if (args[i] %in% arg_names) {
    arg_values[args[i]] <- args[i + 1]
  }
}

# Check if any required arguments are missing
if (any(is.na(arg_values))) {
  stop("\n❌ Missing required arguments.\n\nUsage:\n  Rscript script.R -m MinION_run_name -g genotyping_method -d analysis_date -a analysis_dr\n\nArguments:\n  -m  MinION run name          (e.g. GAMP_22906B2)\n  -g  Genotyping method        (e.g. clair3_diploid)\n  -d  Analysis date            (format: YYYYMMDD, e.g. 20250511)\n  -a  Path to output directory (e.g. /mnt/c/.../output/)\n")
}

# Assign parsed values to variables
MinION_run_name <- arg_values["-m"]
genotyping_method <- arg_values["-g"]
analysis_date <- arg_values["-d"]
analysis_dr <- arg_values["-a"]

# Display parsed values
cat("✅ Parsed arguments:\n")
cat("MinION Run Name:", MinION_run_name, "\n")
cat("Genotyping Method:", genotyping_method, "\n")
cat("Analysis Date:", analysis_date, "\n")
cat("Analysis Directory:", paste0(analysis_dr,'/',MinION_run_name,'/',genotyping_method), "\n")


### Load packages
library("readxl")
library("tidyverse")


# Output path
analysis_dr <- paste0(analysis_dr,'/',MinION_run_name,'/',genotyping_method)

### Output filename
plot_cov_by_amplicon_all_fn <- paste0(analysis_dr, "/plots", "/",MinION_run_name,"_",genotyping_method,"_coverage_plot_",analysis_date,".png")

### Import data
## Coverage for each sample within each MinION run
# define paths
run_cov_fn <- paste0(analysis_dr,"/coverage/",MinION_run_name,"_",genotyping_method,"_coverage_by_run_sample",analysis_date,".csv")

# import data
cov_data <- read.csv(run_cov_fn)

### Annotations for sample type
# Highlight controls
cov_data$is_control <-
  ifelse( (cov_data$sample_id=="Control_Dd2" |
             cov_data$sample_id=="Control_KH2" |
             cov_data$sample_id=="Control_HB3"), "TRUE", "FALSE")

head(cov_data)
cov_data %>% count(is_control)

colnames(cov_data)
cols_to_use <- c("sample_id",
                 "is_control",
                 "crt_coverage_median",
                 "dhfr_coverage_median",
                 "dhps_coverage_median",
                 "mdr1_coverage_median",
                 "k13_coverage_median",
                 "csp_coverage_median")

cov_data_v2 <- cov_data %>%
  select(all_of(cols_to_use))

head(cov_data_v2)


###### Define sample sets
cov_data_all_samps <- cov_data_v2

cov_data_no_controls <- cov_data_v2 %>%
  filter(is_control=="FALSE")

dim(cov_data_all_samps)
dim(cov_data_no_controls)


##### Prepare for plotting
## Prepare the plot features
p_order <- c("crt",
              "dhfr",
              "dhps",
              "mdr1",
              "kelch13",
              "csp")

p_labels <-  c("crt",
               "dhfr",
               "dhps",
               "mdr1",
               "kelch13",
               "csp")


## Prepare data for plotting
cov_data_no_controls_plt <- cov_data_no_controls %>%
  pivot_longer(
    cols = ends_with("coverage_median"),
    values_to = "coverage",
    names_to = "gene_target") %>%
  mutate(gene_target = str_replace(gene_target, "crt_coverage_median", "crt")) %>%
  mutate(gene_target = str_replace(gene_target, "dhfr_coverage_median", "dhfr")) %>%
  mutate(gene_target = str_replace(gene_target, "dhps_coverage_median", "dhps")) %>%
  mutate(gene_target = str_replace(gene_target, "mdr1_coverage_median", "mdr1")) %>%
  mutate(gene_target = str_replace(gene_target, "k13_coverage_median", "kelch13")) %>%
  mutate(gene_target = str_replace(gene_target, "csp_coverage_median", "csp"))

##### Stats

cov_summ_crt <- cov_data_no_controls_plt %>%
  filter(gene_target=="crt")

head(cov_summ_crt)

crt_med_cov <- median(cov_summ_crt$coverage, na.rm=T)
dhfr_med_cov <- median(cov_data_no_controls$dhfr_coverage_median)
dhps_med_cov <- median(cov_data_no_controls$dhps_coverage_median)
mdr1_med_cov <- median(cov_data_no_controls$mdr1_coverage_median)
k13_med_cov <- median(cov_data_no_controls$k13_coverage_median)
csp_med_cov <- median(cov_data_no_controls$csp_coverage_median)

print(paste0("Median coverage for crt: ", crt_med_cov))
print(paste0("Median coverage for dhfr: ", dhfr_med_cov))
print(paste0("Median coverage for dhps: ", dhps_med_cov))
print(paste0("Median coverage for mdr1: ", mdr1_med_cov))
print(paste0("Median coverage for k13: ", k13_med_cov))
print(paste0("Median coverage for csp: ", csp_med_cov))

##### Produce plots
##Excluding controls
p1 <- cov_data_no_controls_plt %>%
  ggplot( aes(x=gene_target, y=coverage)) +
  geom_boxplot(outlier.shape = NA, fill="#e0e0e0") +
  theme_classic() +
  xlab("") +
  ylab("Coverage depth") +
  scale_y_continuous(limits = c(0,20000), expand = c(0, 0)) +
  theme(axis.text.x=element_text(size=11),
        axis.text.y=element_text(size=10),
        axis.title=element_text(size=11),
        axis.text=element_text(colour="black")) +
  geom_jitter(width = 0.1, height = 0, size=1.3, alpha=0.6) +
  scale_x_discrete(limits=p_order,labels=p_labels) +
  theme(plot.margin=unit(c(0.6,0.4,0,0.4),"cm")) # t, r, b, l
p1


#### Save plots
# Ensure the directory exists
dir.create(dirname(plot_cov_by_amplicon_all_fn), recursive = TRUE, showWarnings = FALSE)
ggsave(plot=p1, file=plot_cov_by_amplicon_all_fn, dpi=300, height=4, width=7, units="in")

  

File: analysis/scripts/R3_script_for_coverage_plot_MinION_run_B2.py
import os
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import argparse

# Argument parser setup
parser = argparse.ArgumentParser(
    description="Process amplicon coverage from NanoRAVE Clair3 diploid output"
)

parser.add_argument("-m", "--minion_run_name", required=True, help="MinION run name (e.g., csp_setA_18aprl2025)")
parser.add_argument("-g", "--genotyping_method", required=True, help="Genotyping method (e.g., clair3_diploid)")
parser.add_argument("-d", "--analysis_date", required=True, help="Analysis date (format: YYYYMMDD)")
parser.add_argument("-a", "--analysis_dr", required=True, help="Path to the output directory")

args = parser.parse_args()

# Assign variables
MinION_run_name = args.minion_run_name
genotyping_method = args.genotyping_method
analysis_date = args.analysis_date
analysis_dr = args.analysis_dr

print("📦 Running coverage script with:")
print(f"   MinION run name     : {MinION_run_name}")
print(f"   Genotyping method   : {genotyping_method}")
print(f"   Analysis date       : {analysis_date}")
print(f"   Output directory    : {analysis_dr}")

# Define paths
analysis_path = f"{analysis_dr}/{MinION_run_name}/{genotyping_method}"
coverage_file = f"{analysis_path}/coverage/{MinION_run_name}_{genotyping_method}_coverage_by_run_sample{analysis_date}.csv"
output_plot = f"{analysis_path}/plots/{MinION_run_name}_{genotyping_method}_coverage_plot_{analysis_date}.png"

# Ensure plot directory exists
os.makedirs(os.path.dirname(output_plot), exist_ok=True)

# Load coverage data
cov_data = pd.read_csv(coverage_file)

# Annotate controls
control_samples = ["Control_Dd2", "Control_KH2", "Control_HB3"]
cov_data["is_control"] = cov_data["sample_id"].apply(lambda x: "TRUE" if x in control_samples else "FALSE")

# Identify all gene coverage columns dynamically
coverage_cols = [col for col in cov_data.columns if col.startswith("coverage_median_")]

if not coverage_cols:
    raise ValueError("No coverage_median_* columns found in the input file.")

# Melt the dataframe for seaborn
cov_long = cov_data.melt(id_vars=["sample_id", "is_control"], value_vars=coverage_cols,
                         var_name="gene_target", value_name="coverage")

# Clean gene target names
cov_long["gene_target"] = cov_long["gene_target"].str.replace("coverage_median_", "", regex=False)

# Exclude controls
cov_plot_data = cov_long[cov_long["is_control"] == "FALSE"]

# Plot configuration
sns.set_style("ticks")  # removes grid
plt.figure(figsize=(10, 6))

# Boxplot
sns.boxplot(x="gene_target", y="coverage", data=cov_plot_data, palette="Blues", showfliers=False)

# Jitter plot
sns.stripplot(x="gene_target", y="coverage", data=cov_plot_data, color="black", alpha=0.6, jitter=True)

# Labels & custom styling
plt.xlabel("\nGene Target", fontsize=16)
plt.ylabel("Coverage Depth\n", fontsize=16)
#plt.title("Coverage Depth per Gene") #, fontsize=14)
plt.ylim(0)
plt.yticks(fontsize=14)
plt.xticks(fontsize=14)
sns.despine(top=True, right=True)
plt.tight_layout()

# Save the plot
plt.savefig(output_plot, dpi=300)
print(f"✅ Plot saved at: {output_plot}")

# Show the plot
plt.show()
